e2e/LongHaul/module/IotHub.cs

```diff
@@ -259,6 +261,58 @@ public async Task SetPropertiesAsync(string keyName, object properties, Logger l
             }
         }
 
+        public async Task InvokeDirectMethodOnLeafClientThroughEdgeAsync(string deviceId, string moduleId, string methodName, Logger logger, CancellationToken ct)
+        {
+            Debug.Assert(_moduleClient != null);
+            Debug.Assert(deviceId != null);
+            Debug.Assert(moduleId != null);
+
+            var sw = new Stopwatch();
+            while (!ct.IsCancellationRequested)
+            {
+                try
+                {
+                    var payload = new CustomDirectMethodPayload
+                    {
+                        RandomId = Guid.NewGuid(),
+                        CurrentTimeUtc = DateTimeOffset.UtcNow,
+                        MethodCallsCount = ++_totalMethodCallsModuleToModuleSentCount,
+                    };
+
+                    var directMethodRequest = new EdgeModuleDirectMethodRequest(methodName, payload)
+                    {
+                        ResponseTimeout = TimeSpan.FromSeconds(30),
+                    };
+
+                    logger.Trace($"Invoking direct method for edge device: {deviceId}, edge module: {moduleId}", TraceSeverity.Information);
+                    logger.Metric(TotalDirectMethodCallsModuleToModuleSentCount, _totalMethodCallsModuleToModuleSentCount);
+
+                    sw.Restart();
+
+                    // Invoke the direct method asynchronously and get the response from the simulated module.
+                    DirectMethodResponse response = await _moduleClient
+                        .InvokeMethodAsync(deviceId, moduleId, directMethodRequest, ct)
+                        .ConfigureAwait(false);
+                    logger.Metric(DirectMethodModuleToModuleRoundTripSeconds, sw.Elapsed.TotalSeconds);
+
+                    if (response.TryGetPayload(out CustomDirectMethodPayload responsePayload))
+                    {
+                        logger.Metric(
+                            DirectMethodModuleToModuleDelaySeconds,
+                            (DateTimeOffset.UtcNow - responsePayload.CurrentTimeUtc).TotalSeconds);
+                    }
+
+                    logger.Trace($"Response status: {response.Status}, method name:{methodName}", TraceSeverity.Information);
+                    break;
+                }
+                catch (Exception ex)
+                {
+                    logger.Trace($"Exception invoking direct method from an edge module\n{ex}", TraceSeverity.Warning);
+                    await Task.Delay(s_retryInterval, ct).ConfigureAwait(false);
```

Comment:
Delay should be after the catch so both happy and error paths wait before performing the operation?